feat: Add stream upload (multi-part upload)

wjones127 opened the pull request:
Implements stream / multi-part upload for all object stores except GCP. Exposed as two new methods in `ObjectStore` trait:

```rust
   /// Get a multi-part upload that allows writing data in chunks
    async fn put_multipart(
        &self,
        location: &Path,
    ) -> Result<(MultipartId, Box<dyn AsyncWrite + Unpin + Send>)>;

    /// Cleanup an aborted upload.
    async fn cleanup_multipart(&self, location: &Path, multipart_id: &MultipartId) -> Result<()>;
```

Follow-up work:
 * Need to expose concurrency options for cloud multi-part uploads and set tested default values.

tustvold reviewed (commented):
Awesome that you're working on this :tada: This will open up all sorts of cool things, like doing streaming parquet writes to object storage.

That being said I'm not sure that `BoxStream` will give us the optimal interface. The main challenge with streams is that they are pull-based, which can complicate things like offloading CPU bound work to a separate threadpool.

One option might be to use [futures::Sink](https://docs.rs/futures/latest/futures/sink/trait.Sink.html) i.e. something like

```
type ChunkWriter<'a> = Pin<Box<dyn Sink<Bytes, Error=Error>>>

async fn put_chunked(&self, location: &Path) -> Result<ChunkWriter<'_>>
```

But this has two issues:

* It isn't clear how to handle cancellation
* It still forces data to be produced serially

I wonder if instead we might do something like

```
#[async_trait]
trait ChunkWriter {
  /// Write a chunk at index `idx`
  ///
  /// Takes a shared reference so can be performed in parallel
  async fn put(&self, idx: usize, data: Bytes) -> Result<()>;
  
  /// Abort the multipart upload
  ///
  /// Takes a mutable reference to ensure no inflight requests
  async fn abort(&mut self) -> Result<()>

  /// Finish the multipart upload
  ///
  /// Takes a mutable reference to ensure no inflight requests  
  async fn finish(&mut self) -> Result<()>
}

#[async_trait]
trait ObjectStore {
  async fn put_chunked(&self) -> Result<Box<dyn ChunkWriter + '_>>
  ...
}
```

We could then perhaps `impl Sink for ChunkWriter` as a usability enhancement for simple use-cases. This would let people do

```
store.put_chunked(location)?.await.forward(stream).await?
```

What do you think?

tustvold commented:
> In AWS S3, chunks have to be at least 5 MB

Aah yes, yeah that would be annoying for users to have to worry about

> Apologies if this is a naive question, but are there file format serializers that don't do this?

You could theoretically encode parquet row groups separately, potentially in parallel, or chunks of a CSV or ndjson file, etc... Although tbh uploading them out of order is probably of only limited benefit, especially if internally the `ChunkWriter` can handle multiple concurrent uploads.

> Possibly using something like StreamExt::for_each_concurrent()

I suspect FuturesUnordered or similar might be easier to use in a push-based context.

> So far, I'm more inclined to go with just Sink

Sounds good to me :+1:

> plus some additional trait to handling aborting

Thinking a bit more on this:

* We need to handle uploads aborted by panics, process shutdown, etc...
* Methods like `https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.forward` will drop the sink on error
* We may want to be able to resume uploads

As such I wonder if we need multipart uploads to be a higher level concept on the trait, maybe something like

```
type MultipartWriter<'a> = Pin<Box<dyn Sink<Bytes, Error=Error>>>

pub struct MultipartId(String);

pub struct MultiPart {
  id: MultipartId,
  pub parts: Vec<Part>,
}

pub struct Part {
  pub size: usize,
}

[async_trait]
trait ObjectStore {

  async fn create_multipart(&self, location: &Path) -> Result<MultipartId>;

  async fn write_multipart(&self, id: &MultipartId) -> Result<MultipartWriter<'a>>;  

  async fn abort_multipart(&self, id: &MultipartId) -> Result<()>;

  async fn list_multipart(&self) -> Result<Vec<MultipartId>>;

  async fn get_multipart(&self, id: &MultipartId) -> Result<Multipart>;  
}
```

What do you think?

wjones127 commented:
I've been away from this for a bit, but got back into it this weekend.

I decided to go ahead with the implementation to get a better sense of what the low-level details looks like. I realized that `AsyncWrite` is easier to implement and probably more aligned with what we actually want than `Sink`, so I implemented that.

I learned Azure does *not* provide a way to clean up parts ("blocks" as they call them). There is no unique ID associated with a particular upload, other than the object location. They just expire in 7 days. So abort does nothing for now.
 

> Thinking a bit more on this:
> 
>   1. We need to handle uploads aborted by panics, process shutdown, etc...
>   2. Methods like [Stream::forward](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.forward) will drop the sink on error
>   3. We may want to be able to resume uploads

1. I'm not sure what we can do about this in our error handling. Are you suggesting that we just need to make sure we allow the user to get the `MultipartId` in case they need to abort manually after a crash? Or is there more we can do?
2. I see so that will take ownership of the `Sink`. We wont use that anymore since I implemented `AsyncWrite` instead, but that means there is a problem if there are functions that would consumer `AsyncWrite` because then in the existing design we couldn't abort in case of an error. parquet's [ArrowWriter](https://docs.rs/parquet/16.0.0/parquet/arrow/arrow_writer/struct.ArrowWriter.html) consumes the Writer, so it would probably do the same for an async version. I will refactor the interface then, to make sure we still have some handle to abort from.
3. That would be cool. The Azure case makes this very difficult though, because without a way to clean up past parts, there's no obvious way to differentiate blocks that were part of the upload to resume, or from previously failed upload attempts.

tustvold commented on src/multipart.rs:
```diff
@@ -0,0 +1,197 @@
+        }
+        while let Poll::Ready(Some(res)) = Pin::new(&mut this.tasks).poll_next(cx) {
+            let (part_idx, part) = res?;
+            this.completed_parts.resize_with(
+                std::cmp::max(part_idx + 1, this.completed_parts.len()),
+                || None,
+            );
```
```suggestion
            this.completed_parts.resize(
                std::cmp::max(part_idx + 1, this.completed_parts.len()),
                None,
            );
```

tustvold commented on src/multipart.rs:
```diff
@@ -0,0 +1,197 @@
+            this.current_buffer.extend_from_slice(buf);
+
+            let mut cleared_buffer: Vec<u8> = Vec::new();
+            std::mem::swap(this.current_buffer, &mut cleared_buffer);
```
Could use std::mem::take instead

tustvold commented on src/multipart.rs:
```diff
@@ -0,0 +1,197 @@
+            Poll::Ready(res) => res?,
+        };
+
+        if self.completion_task.is_none() {
```
https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert ?

tustvold commented on src/local.rs:
```diff
@@ -453,6 +455,82 @@ impl ObjectStore for LocalFileSystem {
+        buf: &[u8],
+    ) -> std::task::Poll<Result<usize, io::Error>> {
+        loop {
+            match &mut self.state {
```
I think this could be simplified with maybe_spawn_blocking and `BoxFuture` instead of `JoinHandle`

wjones127 replied:
Main reason I didn't use `maybe_spawn_blocking` was it forces the wrong error type for this function, and don't see any easy way to make that generic. But I agree `BoxFuture` would be better.

tustvold reviewed (commented):
This is looking really nice, mostly just minor nits :+1:

wjones127 commented on src/lib.rs:
```diff
@@ -273,6 +289,11 @@ pub enum Error {
     #[snafu(display("Operation not yet implemented."))]
     NotImplemented,
+
+    #[snafu(display("Error occurred upstream: {}", source))]
```
Ah that's dead code from when I was thinking about accepting a Stream as input to the upload. Needed something to hold errors from the stream. Deleting now.

wjones127 commented on src/memory.rs:
```diff
@@ -208,14 +225,82 @@ impl InMemory {
+        self: Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), io::Error>> {
+        Poll::Ready(Ok(()))
```
I think we *could* do that. Though doing the write in `poll_shutdown` aligns better with the behavior of multi-part upload, since we don't commit the parts until the shutdown method there. In local filesystem, we do flush partial results to the file, so that's the one case where the old value isn't readable during the upload. So I'm inclined to keep final write / commit of data in `poll_shutdown`, but let me know if you still feed differently.

wjones127 commented on src/multipart.rs:
```diff
@@ -0,0 +1,197 @@
+    pub content_id: String,
+}
+
+#[pin_project]
```
Ah turns out it isn't anymore. Was struggling with lifetimes in an earlier iteration. Thanks for pointing that out!

wjones127 commented on src/memory.rs:
```diff
@@ -208,14 +225,82 @@ impl InMemory {
+                    }
+                    Err(_) => {
+                        self.storage
+                            .blocking_write()
```
I've rebased to the recent changes and changed to not use async at all here.

wjones127 commented on src/local.rs:
```diff
@@ -453,6 +455,82 @@ impl ObjectStore for LocalFileSystem {
+        self: Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), io::Error>> {
+        Poll::Ready(Ok(()))
```
So `poll_write()` will return pending until the inner future is totally complete, so I think there shouldn't be any in-flight writes occurring when flush is called.

wjones127 commented on src/local.rs:
```diff
@@ -453,6 +455,82 @@ impl ObjectStore for LocalFileSystem {
     }
 }
 
+enum LocalUploadOperation {
```
Originally I was thinking of having multiple operations we might be waiting for. For example if I was deferring to the inner `write()` and `flush()`. But for simplicity I went with `write_all()`, so I don't actually need it right now. Removed.

wjones127 commented:
> I think put_chunked or put_multi_part might be a better name

"multi-part" seems to be an S3 specific concept, though it also is a decent description. Right now the difference between `put()` and (what is currently called) `upload()` is that the former uploads a single buffer while the latter can upload a stream of buffers. But IIUC we could have `put()` take a stream of buffers too. Then the difference is the `upload()` method would support concurrent uploads for higher throughput on some object stores. So maybe `put_parallel()` or `put_concurrent()`? I will think about this some more...

alamb commented:
@wjones127  and @tustvold -- 

Given we plan to donate the `object_store` crate's code into the arrow-rs repo soon (TM)  see https://github.com/apache/arrow-rs/issues/2030, what is the plan for this PR? Will we merge it here prior to the donation or do you plan to retarget the changes to arrow-rs once it is there?

wjones127 commented:
How soon are we planning to move the crate? I will add the GCS implementation to the PR this weekend, but I can also hold off on that if we want to get this is sooner. I *think* this should be ready to merge in the next week, but feel free to let me know if you believe otherwise @tustvold 😄

alamb commented:
> How soon are we planning to move the crate?

I was imagining sometime next week -- I think we can have a race to see what gets merged first, and the worst case is I'll port this PR's code if needed.

wjones127 commented:
We won't be able to test GCS XML multipart uploads in CI until it's added upstream: https://github.com/fsouza/fake-gcs-server/issues/852

But I can test manually for now.

wjones127 commented:
GCS tests pass against real GCS, but are very slow (takes 2 minutes!). I'll be looking into why that is.

wjones127 commented on src/gcp.rs:
```diff
@@ -122,9 +147,39 @@ struct Object {
+/// Configuration for connecting to [Google Cloud Storage](https://cloud.google.com/storage/).s
 #[derive(Debug)]
 pub struct GoogleCloudStorage {
+    client: Arc<GoogleCloudStorageClient>,
```
I needed to have a reference to the underlying token function and client from the mutlipart writer struct, so I moved the request data into another struct and referenced with an Arc.

tustvold commented on src/multipart.rs:
```diff
@@ -0,0 +1,172 @@
+            return Ok(());
+        }
+        let total_parts = self.completed_parts.len();
+        while let Poll::Ready(Some(res)) = self.tasks.poll_next_unpin(cx) {
```
Technically I think this should be checking https://docs.rs/futures/latest/futures/stream/trait.FusedStream.html#tymethod.is_terminated to ensure it isn't polling a stream that has already yielded `None`

wjones127 replied:
Looking at the source code of FuturesUnordered, I don't see any bad consequences of running `self.tasks.poll_next()` without checking if it's terminated. It should just keep returning `None`.

And it's quite possible that we will return `None` after flushing some data, then add new request futures to the queue and start polling those again.

So I'm not sure if adding such a check will add anything. But let me know if you see a flaw in my understanding.

tustvold replied:
As I said, I think it is probably fine, but technically polling a stream after it has returned None is not well defined (at the trait level).

tustvold commented on src/multipart.rs:
```diff
@@ -0,0 +1,172 @@
+        buf: &[u8],
+    ) -> std::task::Poll<Result<usize, io::Error>> {
+        // Poll current tasks
+        self.as_mut().poll_tasks(cx)?;
```
I had no idea how this was working, turns out Rust 1.36 added some magic to make this work

https://doc.rust-lang.org/src/core/task/poll.rs.html#245

tustvold commented on src/throttle.rs:
```diff
@@ -234,14 +259,93 @@ fn usize_to_u32_saturate(x: usize) -> u32 {
 
+enum ThrottledUploadState {
+    Idle,
+    Sleeping(Pin<Box<dyn Future<Output = ()> + Send>>),
```
```suggestion
    Sleeping(BoxFuture<'static, ()>),
```

Maybe

tustvold commented on src/throttle.rs:
```diff
@@ -234,14 +259,93 @@ fn usize_to_u32_saturate(x: usize) -> u32 {
+    Waiting,
+}
+
+struct ThrottledUpload {
```
As far as I can tell this simply delays the first write or flush, I'm not really sure this is particularly meaningful. As this cannot delay the underlying API calls (as they're hidden behind the AsyncWrite abstraction), I would personally be inclined to just not support these APIs in the ThrottledStore

wjones127 replied:
Is `ThrottledStore` mostly used just for testing?

tustvold replied:
Solely used for testing

wjones127 replied:
I've removed the ThrottleStore implementation.

tustvold reviewed (commented):
This is looking extremely cool, just some relatively minor suggestions

wjones127 commented on Cargo.toml:
```diff
@@ -27,6 +27,7 @@ chrono = { version = "0.4", default-features = false, features = ["clock"] }
 futures = "0.3"
 serde = { version = "1.0", default-features = false, features = ["derive"], optional = true }
 serde_json = { version = "1.0", default-features = false, optional = true }
+serde-xml-rs = { version = "0.5.1", default-features = false, optional = true }
```
Good point. `quick-xml` does look like it might be a better choice. I will try switching to that.

wjones127 replied:
It's a little better, but there is one issue they will not fix that we have to work around: https://github.com/tafia/quick-xml/issues/350

wjones127 commented on src/lib.rs:
```diff
@@ -51,16 +53,41 @@ use snafu::Snafu;
+    ///
+    /// See documentation for individual stores for exact behavior, as capabilities
+    /// vary by object store.
+    async fn cleanup_multipart(&self, location: &Path, multipart_id: &MultipartId) -> Result<()>;
```
Sure. I felt at the time that "cleanup" better communicated that this might be a no-op than "abort", but I don't feel strongly about that now.

tustvold replied:
I think I would prefer abort, otherwise a user might think they need to call cleanup_multipart after a successful upload.

wjones127 commented on src/lib.rs:
```diff
@@ -472,6 +500,68 @@ mod tests {
+        let (upload_id, writer) = storage.put_multipart(&location).await?;
+        drop(writer);
+        storage.cleanup_multipart(&location, &upload_id).await?;
+        let get_res = storage.get(&location).await;
```
Sure; the current implementation of local object store was the only reason I didn't.

wjones127 commented on src/lib.rs:
```diff
@@ -472,6 +500,68 @@ mod tests {
+            get_res.unwrap_err(),
+            crate::Error::NotFound { .. }
+        ));
+
```
It sounds like they get unique upload ids and both should succeed, with one taking precedence based on when. And if object versioning is on, both versions will exist in history. So no error.

https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html

tustvold replied:
Shame

wjones127 commented on src/local.rs:
```diff
@@ -241,6 +231,33 @@ impl ObjectStore for LocalFileSystem {
+    ) -> Result<(MultipartId, Box<dyn AsyncWrite + Unpin + Send>)> {
+        let path = self.config.path_to_filesystem(location)?;
+
+        let file = open_writable_file(&path)?;
```
That's a nice idea! I'll try that.

wjones127 commented on src/multipart.rs:
```diff
@@ -0,0 +1,172 @@
+
+#[derive(Debug, Clone)]
+pub(crate) struct UploadPart {
+    pub content_id: String,
```
content_id is sometimes the ETag and sometimes the part number, depending on what the underlying implementation wants, so probably shouldn't call it `part_id`. I can add that explanation in a comment.

wjones127 commented on src/multipart.rs:
```diff
@@ -0,0 +1,172 @@
+        &self,
+        buf: Vec<u8>,
+        part_idx: usize,
+    ) -> BoxFuture<'static, Result<(usize, UploadPart), io::Error>>;
```
More consistent maybe, but feels like unnecessary work to me (both programmer and CPU).

tustvold reviewed (approved):
Awesome :+1:
